### ISSUE TITLE
"constant expression type mismatch" for valid code

### ISSUE THREAD

**Issue by dev-zero** [flang-compiler/flang#424]
given the following two files

`dbcsr_data_types.f90`:
```fortran
module dbcsr_data_types
   TYPE dbcsr_data_obj
      TYPE(dbcsr_data_area_type), POINTER           :: d => Null()
   END TYPE dbcsr_data_obj

   TYPE dbcsr_mempool_type
      TYPE(dbcsr_mempool_entry_type), POINTER :: root => Null()
      INTEGER                                 :: capacity = 1
   END TYPE dbcsr_mempool_type

   TYPE dbcsr_mempool_entry_type
      TYPE(dbcsr_data_obj)                    :: area
      TYPE(dbcsr_mempool_entry_type), POINTER :: next => Null()
   END TYPE dbcsr_mempool_entry_type

   TYPE dbcsr_memtype_type
      TYPE(dbcsr_mempool_type), POINTER :: pool => Null()
   END TYPE dbcsr_memtype_type

   TYPE(dbcsr_memtype_type), PARAMETER :: dbcsr_memtype_default = dbcsr_memtype_type(pool=Null())

   TYPE dbcsr_data_area_type
      INTEGER                                  :: ref_size = -1
      INTEGER                                  :: refcount = -1
      INTEGER                                  :: id = -1
      TYPE(dbcsr_memtype_type)                 :: memory_type = dbcsr_memtype_default
      INTEGER                                  :: data_type = -1
   END TYPE dbcsr_data_area_type

end module
```

`dbcsr_data_types_user.f90`:
```fortran
module a
   use dbcsr_data_types, only: dbcsr_memtype_type

   type foo
      type(dbcsr_memtype_type) :: val = dbcsr_memtype_type()
   end type
end module
```

trying to compile them using

```bash
flang -c dbcsr_data_types.f90
flang -c dbcsr_data_types_user.f90
```

gives

```
/tmp/dbcsr_data_types_user-985010.ll:9:47: error: constant expression type mismatch
@_a_10_ = global %struct_a_10_ < { [16 x i8]  [i8 0,i8 0,i8 0,i8 0,i8 0,i8 0,i8 0,i8 0]  } > , align 64
                                              ^
1 error generated.
```
for `dbcsr_data_types_user.f90`.

Works with gfortran-4.8 and gfortran-8.

**Comment by pawosm-arm** [comment]
I tried some crazy hack that emits zeroinitializer instead of calling put_skip() for emitting stream of "i8 0"'s , seems to solve this issue.

**Comment by pawosm-arm** [comment]
Sadly, fixing this issue does not make cp2k buildable, still fix for #239 is needed.

**Comment by pawosm-arm** [comment]
But... it's not how my commit looked like! I guess it was malformed by changes made by further commit to the code around. Corrections are introduced by a new pull-request #561

**Comment by pawosm-arm** [comment]
I guess it's time to pull the request and close this one.

**Comment by sscalpone** [comment]
Working on it....  Waiting until the big C++ pull is finished.

**Comment by gklimowicz** [comment]
Paul: Does this still need the pull request #561 to satisfy this?

**Comment by pawosm-arm** [comment]
@gklimowicz yes, I've made a mistake preparing initial pull request, unfortunately, before I noticed that, it was already accepted. #561 corrects the change that was pulled in and fixes the initial issue #424

**Comment by gklimowicz** [comment]
@pawosm-arm Thanks. I will close this when PR #561 is handled.
